imagick/kernel_info.go

```diff
@@ -6,9 +6,145 @@ package imagick
 
 /*
 #include <magick/MagickCore.h>
+#include <magick/geometry.h>
+#include <magick/morphology.h>
 */
 import "C"
 
+//import "runtime"
+import "unsafe"
+
+// This struct represents the KernelInfo C API of ImageMagick
 type KernelInfo struct {
 	info *C.KernelInfo
 }
+
+func newKernelInfo(cki *C.KernelInfo) *KernelInfo {
+	ki := &KernelInfo{info: cki}
+	//TODO - understand and implement this.
+	// runtime.SetFinalizer(ki, Destroy)
+	//ki.IncreaseCount()
+	return ki
+}
+
+// Convert the current KernelInfo to an 2d-array of values. The values are either
+// a float64 if the element is used, or NaN if the element is not used by the kernel
+func (kernel_info *KernelInfo) ToArray() [][]float64 {
+	count := 0
+	values := [][]float64{}
+
+	for y := C.size_t(0); y < kernel_info.info.height; y++ {
+		row_values := make([]float64, kernel_info.info.width)
+		for x := C.size_t(0); x < kernel_info.info.width; x++ {
+			p2 := (*[1 << 10]C.double)(unsafe.Pointer(kernel_info.info.values))
+			row_values[x] = float64(p2[count])
+			count++
+		}
+		values = append(values, row_values)
+	}
+
+	return values
+}
+
+// Create a kernel from a builtin in kernel. See http://www.imagemagick.org/Usage/morphology/#kernel
+// for examples. Currently the 'rotation' symbols are not supported. Example:
+// kernel_info := imagick.AcquireKernelBuiltIn(imagick.KERNEL_RING, "2,1")
+func AcquireKernelBuiltIn(kernel_type KernelInfoType, kernelString string) *KernelInfo {
+	geometry_info := (*GeometryInfo)(C.malloc(C.size_t(unsafe.Sizeof(C.GeometryInfo{}))))
```

Comment:
I don't fully understand this part. It mallocs (with no free) a `C.GeometryInfo` and casts it to a pointer to a Go struct? It doesn't seem the `C.GeometryInfo` is even needed outside this function, is that right? So do we even need the go struct for it or the malloc? Could we just create a value type on the stack, and fill it with the data and then pass it off to the functions that need it?